sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -735,70 +735,100 @@ class CodegenContext {
   }
 
   /**
-   * Generates code creating a [[UnsafeArrayData]].
+   * Generates code creating a [[UnsafeArrayData]] or [[GenericArrayData]] based on
+   * given parameters.
    *
    * @param arrayName name of the array to create
+   * @param elementType data type of the elements in source array
    * @param numElements code representing the number of elements the array should contain
-   * @param elementType data type of the elements in the array
    * @param additionalErrorMessage string to include in the error message
+   * @param elementSize optional value which shows the size of an element of the allocated
+   *                    [[UnsafeArrayData]] or [[GenericArrayData]]
+   *
+   * @return code representing the allocation of [[ArrayData]]
+   *         code representing a setter of an assignment for the generated array
    */
-  def createUnsafeArray(
+  def createArrayData(
       arrayName: String,
-      numElements: String,
       elementType: DataType,
-      additionalErrorMessage: String): String = {
-    val arraySize = freshName("size")
-    val arrayBytes = freshName("arrayBytes")
+      numElements: String,
+      additionalErrorMessage: String,
+      elementSize: Option[Int] = None): (String, String) = {
```

Comment:
Yeah, they are exclusive. However, when `elementType` is not used, the integer value is required.

`MapEntries` stores `UnsafeRow` instead of primitive data. [This](https://github.com/apache/spark/blob/c15630014c7ef850fa21c1247fa5edd2b1bac81b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala#L405) requires non-fixed size unlike `elementType.defaultSize`.